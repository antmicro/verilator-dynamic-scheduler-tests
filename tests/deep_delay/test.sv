module t;
    // As Verilator doesn't support recursive calls, let's use macros to
    // generate functions for a deep call stack
    `define DEEP_DELAY_END(i) \
    task deep_delay``i; \
        $write("[%0t] at depth %0d\n", $time, i); \
    endtask

    `define DEEP_DELAY(i, j) \
    task deep_delay``i; \
        $write("[%0t] entering depth %0d\n", $time, i); \
        #1 deep_delay``j(); \
        #1 $write("[%0t] leaving depth %0d\n", $time, i); \
    endtask

    `DEEP_DELAY_END(10);
    `DEEP_DELAY(9, 10);
    `DEEP_DELAY(8, 9);
    `DEEP_DELAY(7, 8);
    `DEEP_DELAY(6, 7);
    `DEEP_DELAY(5, 6);
    `DEEP_DELAY(4, 5);
    `DEEP_DELAY(3, 4);
    `DEEP_DELAY(2, 3);
    `DEEP_DELAY(1, 2);

    initial begin
        deep_delay1;
        $write("*-* All Finished *-*\n");
        $finish;
    end
endmodule
