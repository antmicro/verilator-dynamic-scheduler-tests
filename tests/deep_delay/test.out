[0] entering depth 1
[1] entering depth 2
[2] entering depth 3
[3] entering depth 4
[4] entering depth 5
[5] entering depth 6
[6] entering depth 7
[7] entering depth 8
[8] entering depth 9
[9] at depth 10
[10] leaving depth 9
[11] leaving depth 8
[12] leaving depth 7
[13] leaving depth 6
[14] leaving depth 5
[15] leaving depth 4
[16] leaving depth 3
[17] leaving depth 2
[18] leaving depth 1
*-* All Finished *-*
